proposals/NNNN-swift-distributed-tracing.md

```diff
@@ -0,0 +1,306 @@
+
+Whereas [Logging](https://github.com/apple/swift-log) and [Metrics](https://github.com/apple/swift-metrics) enables
+you to instrument specific parts of an application,
+[Distributed Tracing](https://github.com/apple/swift-distributed-tracing) provides a holistic view on an entire
+distributed system. Together with these two, Distributed Tracing will complete the "three pillars of observability".
+
+As with Logging and Metrics, the community will benefit most from Distributed Tracing if it is implemented in
+libraries and frameworks directly using one common API. End-users should be able to freely choose a suitable backend
+implementation without having to change the libraries or frameworks they're using.
+
+## Proposed Solution
+
+Swift Distributed Tracing revolves around creating **spans**, which together form a tree-like structure. A trace can be
+made up by spans recorded in a single service, or propagated across multiple services. Swift Distributed Tracing uses
+the task-local based [Swift Service Context](https://github.com/apple/swift-service-context) to implement propagation
+transparently, without the need of manual context passing.
+
+Our proposed solution is a library which targets three "roles":
+
+- End-users
+- Library & framework authors
+- Tracer backend implementations
+
+### End-users
+
+End-users are the ones benefiting from Distributed Tracing. They choose a Tracing backend that suits their needs, use
+libraries that have built-in Swift Distributed Tracing support, and augment their own code with manual instrumentation.
+
+### Library & framework authors
+
+Libraries/Frameworks such as HTTP servers/clients, database libraries, etc. know best how to instrument their library's
+internals. They implement generic tracing support using the Swift Distributed Tracing API without a specific tracing
+backend in mind.
+
+#### Examples
+
+- [Hummingbird](https://github.com/hummingbird-project/hummingbird/releases/tag/1.6.0)
+- [Soto](https://github.com/soto-project/soto-core/pull/575)
+
+### Tracer backend implementations
+
+The final piece of the puzzle are tracer backend implementations. They provide vendor-specific support for exporting
+tracing spans.
+
+#### Examples
+
+[Swift OTel](https://github.com/slashmo/swift-otel) exposes a tracer that exports to
+an [OpenTelemetry Collector](https://opentelemetry.io/docs/collector). This already allows adopters of this tracing
+library to export to popular backends compatible with OpenTelemetry, such as Zipkin, Jaeger, Honeycomb, and more.
+
+## Detailed design
+
+Both library/framework authors and end-users use the same API to create tracing spans. Adding a span to functions can
+be achieved by using the `withSpan(operationName:context:body:)` API:
+
+```swift
+func performOperation() -> String {
+    withSpan("operation") {
+        return "handled"
+    }
+}
+```
+
+### Customizing the span
+
+A span can be further customized using `withSpan(operationName:context:body:)` as well:
```

Comment:
How is the span customised using `withSpan(operationName:context:body:)` in the example below. The text implies the `withSpan` function is doing the customisation. Also the function signature does not include `ofKind:`which you use in the example

Reply:
Sorry I don't get this question:
> How is the span customised using withSpan(operationName:context:body:) in the example below.

> ```
>     withSpan(request.endpointPath, ofKind: .server) { span in
>         span.attributes["http.method"] = request.method.rawValue
> ```

It does that ^ but the primary purpose is to set the task-local. Should we explain this more?

Reply:
I added: [`da9f679` (#84)](https://github.com/swift-server/sswg/pull/84/commits/da9f6790f53ed9f4162f173f04cd11b2bb9d04ab)

Reply:
It is just the text doesn't align with the example. You imply it is the function `withSpan` that is doing the customisation, not the editing of the attributes `span.attributes["http.method] = ...`. Its not a huge issue